apps/charterafrica/src/lib/tools/models.js

```diff
@@ -0,0 +1,132 @@
+    }
+    const objKey = obj[key];
+    if (!uniqueKeys.has(objKey)) {
+      uniqueKeys.add(objKey);
+      return true;
+    }
+    return false;
+  });
+}
+
+export const createOrganization = async (toCreate) => {
+  const { docs } = await api.getCollection(ORGANIZATION_COLLECTION, {
+    where: {
+      externalId: { equals: toCreate.externalId },
+    },
+  });
+  if (docs.length) {
+    const data = await api.updateCollection(
+      ORGANIZATION_COLLECTION,
+      docs[0]?.id,
+      {
+        ...toCreate,
+        updatedAt: new Date(),
+      }
+    );
+    return data;
+  }
+  const data = await api.createCollection(ORGANIZATION_COLLECTION, toCreate);
+  return data;
+};
+
+export const createPerson = async (toCreate) => {
+  const { docs } = await api.getCollection(PEOPLE_COLLECTION, {
+    where: {
+      externalId: { equals: toCreate.externalId },
+    },
+  });
+  if (docs.length) {
+    const data = await api.updateCollection(PEOPLE_COLLECTION, docs[0]?.id, {
+      ...toCreate,
+      updatedAt: new Date(),
+    });
+    return data;
+  }
+  const data = await api.createCollection(PEOPLE_COLLECTION, toCreate);
+  return data;
+};
+
+export const bulkCreatePeople = async (contributors = []) => {
+  const promises = contributors?.map(async (contributor) => {
+    return createPerson(contributor);
+  });
+  const results = await Promise.all(promises);
+  return results;
+};
+
+export const bulkCreateOrganisations = async (organisations = []) => {
+  const promises = organisations?.map(async (organisation) => {
+    return createOrganization(organisation);
+  });
+  const results = await Promise.all(promises);
+  return results;
+};
+
+export const updateOrCreateTool = async (data) => {
+  try {
+    const { organisation, people, id, ...rest } = data;
+    const createdOrganization = organisation?.externalId
+      ? await createOrganization(organisation)
+      : null;
+    const createdPeople = await bulkCreatePeople(people);
+    const toCreate = {
+      ...rest,
+      people: createdPeople.map((person) => person?.id),
+      organisation: createdOrganization?.id,
+      updatedAt: new Date(),
+    };
+    const queryArgs = {
+      where: {
+        ...(id ? { id: { equals: id } } : {}),
+        externalId: data?.externalId ? { equals: data?.externalId } : undefined,
+      },
+    };
+    const { docs } = await api.getCollection(TOOL_COLLECTION, queryArgs);
+    if (docs?.length) {
+      if (id) {
+        const res = await api.updateCollection(TOOL_COLLECTION, id, toCreate);
+        return res;
+      }
+      return docs[0];
+    }
+    const res = await api.createCollection(TOOL_COLLECTION, toCreate);
+    return res;
+  } catch (error) {
+    throw new FetchError(error.message, data, 500);
+  }
+};
+
+export const bulkCreateTools = async (tools) => {
+  const organisationsToCreate = removeDuplicatesByKey(
+    tools.map(({ organisation }) => organisation),
+    "externalId"
+  ).filter((org) => !!org.externalId);
```

Comment:
What's happening here?

Reply:
we are removing null enries.

Some tools from github have organisation as `null`

Reply:
Cool. why aren't we filtering them out wherever we're first loading them from GH so that the rest of the code doesn't need to deal with this issue?